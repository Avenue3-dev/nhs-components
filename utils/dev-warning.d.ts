/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
