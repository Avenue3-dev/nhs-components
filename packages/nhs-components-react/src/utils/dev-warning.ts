const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
