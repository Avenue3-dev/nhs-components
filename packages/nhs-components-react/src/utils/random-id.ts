export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
