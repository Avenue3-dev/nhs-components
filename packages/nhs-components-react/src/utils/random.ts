export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
