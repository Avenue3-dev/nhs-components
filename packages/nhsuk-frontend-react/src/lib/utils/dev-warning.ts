const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
