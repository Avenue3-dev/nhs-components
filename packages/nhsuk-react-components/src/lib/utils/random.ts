export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
