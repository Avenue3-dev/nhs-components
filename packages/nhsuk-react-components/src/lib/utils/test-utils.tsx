/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
