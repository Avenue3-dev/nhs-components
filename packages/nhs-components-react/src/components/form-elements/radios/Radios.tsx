import Label from '@/components/core/label/Label'
import clsx from 'clsx'
import {
  ForwardRefRenderFunction,
  HTMLProps,
  ReactNode,
  forwardRef,
  useCallback,
  useEffect,
  useReducer,
  useState,
} from 'react'
import RadiosContext, {
  RadiosContextValue,
  reducer,
  useRadiosContext,
} from './RadiosContext'
import Hint from '@/components/form-elements/hint'
import FormGroup from '@/components/core/form-group/FormGroup'
import { FormElementProps } from '@/types/form-element-types'
import { generateRandomString } from '@/utils/random'

type Radios = {
  Divider: typeof Divider
  Radio: typeof RadioForwardRef
} & React.FC<RadioContainerProps>

type RadioProps = {
  conditional?: ReactNode
  forceShowConditional?: boolean
  conditionalWrapperProps?: HTMLProps<HTMLDivElement>
} & HTMLProps<HTMLInputElement> &
  Omit<
    FormElementProps,
    'error' | 'errorProps' | 'formGroupProps' | 'disableErrorLine'
  >

type RadioContainerProps = {
  inline?: boolean
  idPrefix?: string
} & HTMLProps<HTMLDivElement> &
  FormElementProps

const Divider: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <div className={clsx('nhsuk-radios__divider', className)} {...rest}>
      {children}
    </div>
  )
}

const Radio: ForwardRefRenderFunction<HTMLInputElement, RadioProps> = (
  {
    children,
    className,
    id,
    hint,
    hintProps,
    labelProps,
    conditional,
    forceShowConditional,
    conditionalWrapperProps,
    checked,
    defaultChecked,
    onChange,
    type = 'radio',
    ...rest
  },
  ref,
): JSX.Element => {
  const {
    name,
    getRadioId,
    leaseReference,
    releaseReference,
    selectedRadio,
    dispatch,
  } = useRadiosContext()
  const [radioReference] = useState<string>(leaseReference())
  const inputId = id || getRadioId(radioReference)
  const shouldShowConditional =
    selectedRadio === radioReference && checked !== false

  const { className: labelClassName, ...labelRest } = labelProps || {}
  const { className: hintClassName, ...hintRest } = hintProps || {}
  const { className: conditionalClassName, ...conditionalRest } =
    conditionalWrapperProps || {}

  const setSelected = (reference: string): void => {
    dispatch({ type: 'set_selected', data: { refId: reference } })
  }

  useEffect(() => {
    if (defaultChecked) {
      setSelected(radioReference)
    }

    return () => releaseReference(radioReference)
  }, [])

  useEffect(() => {
    if (checked) {
      setSelected(radioReference)
    }
  }, [checked])

  useEffect(() => {
    dispatch({
      type: 'set_conditional',
      data: { refId: radioReference, hasConditional: Boolean(conditional) },
    })

    return () =>
      dispatch({
        type: 'set_conditional',
        data: { refId: radioReference, hasConditional: false },
      })
  }, [conditional])

  return (
    <>
      <div className={clsx('nhsuk-radios__item', className)}>
        <input
          className="nhsuk-radios__input"
          id={inputId}
          name={name}
          type={type}
          checked={checked}
          defaultChecked={defaultChecked}
          ref={ref}
          onChange={(ev) => {
            setSelected(radioReference)
            onChange?.(ev)
          }}
          {...rest}
        />
        {children && (
          <Label
            className={clsx('nhsuk-radios__label', labelClassName)}
            id={`${inputId}--label`}
            htmlFor={inputId}
            {...labelRest}
          >
            {children}
          </Label>
        )}
        {hint && (
          <Hint
            className={clsx('nhsuk-radios__hint', hintClassName)}
            id={`${inputId}--hint`}
            {...hintRest}
          >
            {hint}
          </Hint>
        )}
      </div>
      {conditional && (shouldShowConditional || forceShowConditional) && (
        <div
          className={clsx('nhsuk-radios__conditional', conditionalClassName)}
          id={`${inputId}--conditional`}
          {...conditionalRest}
        >
          {conditional}
        </div>
      )}
    </>
  )
}

/**
 * Radio component that renders a single radio input.
 *
 * @param {RadioProps} props - The props to be passed to the radio input.
 * @param {React.ForwardedRef<HTMLInputElement>} ref - Optional ref to be passed to the radio input.
 *
 * @example
 * ```tsx
 * <Radios.Radio value="yes" hint="some hint">
 *  Yes
 * </Radios.Radio>
 * ```
 */
const RadioForwardRef = forwardRef(Radio)

/**
 * Use radios when you want users to select only 1 option from a list.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/radios).
 *
 * @example
 * ```tsx
 * <Radios label="Do you have any allergies?">
 *  <Radios.Radio value="yes" hint="some hint">
 *   Yes
 *  </Radios.Radio>
 *  <Radios.Radio value="no">No</Radios.Radio>
 * </Radios>
 * ```
 */
const Radios: Radios = ({ children, ...rest }): JSX.Element => {
  let radioIds: Record<string, string> = {}
  let radioReferences: string[] = []
  let radioCount = 0

  const [state, dispatch] = useReducer(reducer, {
    selected: '',
    conditional: [],
  })

  const getRadioId = useCallback((id: string, reference: string) => {
    if (reference in radioIds) {
      return radioIds[reference]
    }

    radioCount += 1
    radioIds[reference] = `${rest.idPrefix || id}-${radioCount}`
    return radioIds[reference]
  }, [])

  const leaseReference = (): string => {
    const reference = generateRandomString()

    if (radioReferences.includes(reference)) {
      return leaseReference()
    }

    radioReferences.push(reference)
    return reference
  }

  const releaseReference = (reference: string): void => {
    radioReferences = radioReferences.filter((ref) => ref !== reference)
  }

  const resetRadios = (): void => {
    radioIds = {}
    radioCount = 0
  }

  return (
    <FormGroup<RadioContainerProps>
      {...rest}
      inputType="radios"
      render={({ id, name, inline, className, ...renderRest }) => {
        resetRadios()

        const radiosContextValue: RadiosContextValue = {
          name,
          getRadioId: (reference) => getRadioId(id, reference),
          selectedRadio: state.selected,
          leaseReference,
          releaseReference,
          dispatch,
        }

        return (
          <div
            className={clsx(
              'nhsuk-radios',
              { 'nhsuk-radios--inline': inline },
              className,
            )}
            id={id}
            {...renderRest}
          >
            <RadiosContext.Provider value={radiosContextValue}>
              {children}
            </RadiosContext.Provider>
          </div>
        )
      }}
    />
  )
}

Divider.displayName = 'Radios.Divider'
RadioForwardRef.displayName = 'Radios.Radio'
Radios.displayName = 'Radios'

Radios.Divider = Divider
Radios.Radio = RadioForwardRef

export default Radios
