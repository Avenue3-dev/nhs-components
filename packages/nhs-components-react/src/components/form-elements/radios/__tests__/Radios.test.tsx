import { render, screen } from '@/utils/test-utils'
import { describe, test } from 'vitest'
import Radios from '..'

describe('Radios', () => {
  test('should match the snapshot', () => {
    render(
      <Radios data-testid="radios" label="Label">
        <Radios.Radio value="1">Option 1</Radios.Radio>
        <Radios.Radio value="2">Option 2</Radios.Radio>
      </Radios>,
    )

    expect(screen.getByTestId('radios').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - hint, error', () => {
    render(
      <Radios data-testid="radios" label="Label" error="Error">
        <Radios.Radio value="1" hint="Hint">
          Option 1
        </Radios.Radio>
        <Radios.Radio value="2">Option 2</Radios.Radio>
      </Radios>,
    )

    expect(screen.getByTestId('radios').parentNode).toMatchSnapshot()
  })

  test('should match the snapshot - conditional', () => {
    render(
      <Radios data-testid="radios" label="Label" error="Error">
        <Radios.Radio defaultChecked value="1" hint="Hint" conditional="Yay">
          Option 1
        </Radios.Radio>
        <Radios.Radio value="2">Option 2</Radios.Radio>
      </Radios>,
    )

    expect(screen.getByTestId('radios').parentNode).toMatchSnapshot()
  })
})
