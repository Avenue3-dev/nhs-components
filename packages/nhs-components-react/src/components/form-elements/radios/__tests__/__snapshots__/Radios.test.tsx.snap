// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radios > should match the snapshot - conditional 1`] = `
<div
  class="nhsuk-form-group nhsuk-form-group--error"
>
  <label
    class="nhsuk-label"
    for="radios_:r2:"
    id="radios_:r2:--label"
  >
    Label
  </label>
  <span
    class="nhsuk-error-message"
    for="radios_:r2:"
    id="radios_:r2:--error-message"
  >
    <span
      class="nhsuk-u-visually-hidden"
    >
      Error
      :
    </span>
    Error
  </span>
  <div
    aria-labelledby="radios_:r2:--label"
    class="nhsuk-radios"
    data-testid="radios"
    error="Error"
    id="radios_:r2:"
  >
    <div
      class="nhsuk-radios__item"
    >
      <input
        class="nhsuk-radios__input"
        id="radios_:r2:-1"
        name="radios_:r2:"
        type="radio"
        value="1"
      />
      <label
        class="nhsuk-label nhsuk-radios__label"
        for="radios_:r2:-1"
        id="radios_:r2:-1--label"
      >
        Option 1
      </label>
      <div
        class="nhsuk-hint nhsuk-radios__hint"
        id="radios_:r2:-1--hint"
      >
        Hint
      </div>
    </div>
    <div
      class="nhsuk-radios__conditional"
      id="radios_:r2:-1--conditional"
    >
      Yay
    </div>
    <div
      class="nhsuk-radios__item"
    >
      <input
        class="nhsuk-radios__input"
        id="radios_:r2:-2"
        name="radios_:r2:"
        type="radio"
        value="2"
      />
      <label
        class="nhsuk-label nhsuk-radios__label"
        for="radios_:r2:-2"
        id="radios_:r2:-2--label"
      >
        Option 2
      </label>
    </div>
  </div>
</div>
`;

exports[`Radios > should match the snapshot - hint, error 1`] = `
<div
  class="nhsuk-form-group nhsuk-form-group--error"
>
  <label
    class="nhsuk-label"
    for="radios_:r1:"
    id="radios_:r1:--label"
  >
    Label
  </label>
  <span
    class="nhsuk-error-message"
    for="radios_:r1:"
    id="radios_:r1:--error-message"
  >
    <span
      class="nhsuk-u-visually-hidden"
    >
      Error
      :
    </span>
    Error
  </span>
  <div
    aria-labelledby="radios_:r1:--label"
    class="nhsuk-radios"
    data-testid="radios"
    error="Error"
    id="radios_:r1:"
  >
    <div
      class="nhsuk-radios__item"
    >
      <input
        class="nhsuk-radios__input"
        id="radios_:r1:-1"
        name="radios_:r1:"
        type="radio"
        value="1"
      />
      <label
        class="nhsuk-label nhsuk-radios__label"
        for="radios_:r1:-1"
        id="radios_:r1:-1--label"
      >
        Option 1
      </label>
      <div
        class="nhsuk-hint nhsuk-radios__hint"
        id="radios_:r1:-1--hint"
      >
        Hint
      </div>
    </div>
    <div
      class="nhsuk-radios__item"
    >
      <input
        class="nhsuk-radios__input"
        id="radios_:r1:-2"
        name="radios_:r1:"
        type="radio"
        value="2"
      />
      <label
        class="nhsuk-label nhsuk-radios__label"
        for="radios_:r1:-2"
        id="radios_:r1:-2--label"
      >
        Option 2
      </label>
    </div>
  </div>
</div>
`;

exports[`Radios > should match the snapshot 1`] = `
<div
  class="nhsuk-form-group"
>
  <label
    class="nhsuk-label"
    for="radios_:r0:"
    id="radios_:r0:--label"
  >
    Label
  </label>
  <div
    aria-labelledby="radios_:r0:--label"
    class="nhsuk-radios"
    data-testid="radios"
    id="radios_:r0:"
  >
    <div
      class="nhsuk-radios__item"
    >
      <input
        class="nhsuk-radios__input"
        id="radios_:r0:-1"
        name="radios_:r0:"
        type="radio"
        value="1"
      />
      <label
        class="nhsuk-label nhsuk-radios__label"
        for="radios_:r0:-1"
        id="radios_:r0:-1--label"
      >
        Option 1
      </label>
    </div>
    <div
      class="nhsuk-radios__item"
    >
      <input
        class="nhsuk-radios__input"
        id="radios_:r0:-2"
        name="radios_:r0:"
        type="radio"
        value="2"
      />
      <label
        class="nhsuk-label nhsuk-radios__label"
        for="radios_:r0:-2"
        id="radios_:r0:-2--label"
      >
        Option 2
      </label>
    </div>
  </div>
</div>
`;
